tests/fsharp/PlatformHelpers.fs

```diff
@@ -0,0 +1,204 @@
+        if Path.IsPathRooted(exe) then exe
+        else 
+            match Path.GetDirectoryName(exe) with
+            | "" -> exe
+            | _ -> Path.Combine(baseDir,exe) |> Path.GetFullPath
+
+    let exec cmdArgs (workDir: FilePath) envs (path: FilePath) arguments =
+
+        let exePath = path |> processExePath workDir
+        let processInfo = new ProcessStartInfo(exePath, arguments)
+        processInfo.CreateNoWindow <- true
+        processInfo.UseShellExecute <- false
+        processInfo.WorkingDirectory <- workDir
+
+        envs
+        |> Map.iter (fun k v -> processInfo.EnvironmentVariables.[k] <- v)
+
+        let p = new Process()
+        p.EnableRaisingEvents <- true
+        p.StartInfo <- processInfo
+
+        cmdArgs.RedirectOutput
+        |> Option.map (fun f -> (fun (ea: DataReceivedEventArgs) -> ea.Data |> f)) 
+        |> Option.iter (fun newOut ->
+            processInfo.RedirectStandardOutput <- true
+            p.OutputDataReceived.Add newOut
+        )
+
+        cmdArgs.RedirectError 
+        |> Option.map (fun f -> (fun (ea: DataReceivedEventArgs) -> ea.Data |> f)) 
+        |> Option.iter (fun newErr ->
+            processInfo.RedirectStandardError <- true
+            p.ErrorDataReceived.Add newErr
+        )
+
+        cmdArgs.RedirectInput
+        |> Option.iter (fun _ -> p.StartInfo.RedirectStandardInput <- true)
+
+        let exitedAsync (proc: Process) =
+            let tcs = new System.Threading.Tasks.TaskCompletionSource<int>();
+            p.Exited.Add (fun s -> 
+                tcs.TrySetResult(proc.ExitCode) |> ignore
+                proc.Dispose())
+            tcs.Task
+
+        p.Start() |> ignore
+    
+        cmdArgs.RedirectOutput |> Option.iter (fun _ -> p.BeginOutputReadLine())
+        cmdArgs.RedirectError |> Option.iter (fun _ -> p.BeginErrorReadLine())
+
+        cmdArgs.RedirectInput
+        |> Option.map (fun input -> async {
+            let inputWriter = p.StandardInput
+            do! inputWriter.FlushAsync () |> Async.AwaitIAsyncResult |> Async.Ignore
+            input inputWriter
+            do! inputWriter.FlushAsync () |> Async.AwaitIAsyncResult |> Async.Ignore
+            inputWriter.Close ()
+            })
+        |> Option.iter Async.Start
+
+        let exitCode = p |> exitedAsync |> Async.AwaitTask |> Async.RunSynchronously
+
+        match exitCode with
+        | 0 -> Success
+        | err -> ErrorLevel err
+
+
+
+type Result<'S,'F> =
+    | Success of 'S
+    | Failure of 'F
+
+type Attempt<'S,'F> = (unit -> Result<'S,'F>)
+
+open System.Diagnostics
+
+[<DebuggerStepThrough>]
+let internal succeed x = (fun () -> Success x)
+
+[<DebuggerStepThrough>]
+let internal failed err = (fun () -> Failure err)
+
+[<DebuggerStepThrough>]
+let runAttempt (a: Attempt<_,_>) = a ()
+
+[<DebuggerStepThrough>]
+let delay f = (fun () -> f() |> runAttempt)
+
+[<DebuggerStepThrough>]
+let either successTrack failTrack (input : Attempt<_, _>) : Attempt<_, _> =
```

Comment:
I have quite a strong aversion to two-way functionals like ``either`` to complexify what could be just a simple pattern match. Please just remove this and define bind and fail and any other uses via a pattern match. thanks